javascript/extractor/src/com/semmle/js/parser/RegExpParser.java

```diff
@@ -314,9 +314,29 @@ private RegExpTerm parseAtom() {
       return this.finishTerm(new Group(loc, capture, number, name, dis));
     }
 
-    char c = this.nextChar();
-    if ("^$\\.*+?()[]{}|".indexOf(c) != -1) this.error(Error.UNEXPECTED_CHARACTER, this.pos - 1);
-    return this.finishTerm(new Constant(loc, String.valueOf(c)));
+    // Parse consecutive constants into a single Constant node.
+    // Due to speculative parsing of string literals, this part of the code is fairly hot.
+    int startPos = this.pos;
+    int endPos = startPos;
+    while (endPos < src.length()) {
+      if ("^$\\.*+?()[]{}|".indexOf(src.charAt(endPos)) != -1) break;
+      ++endPos;
+    }
+    if (startPos == endPos) {
+      this.error(Error.UNEXPECTED_CHARACTER, endPos);
+      endPos = startPos + 1; // To ensure progress, make sure we parse at least one character.
+    }
+    if (endPos != startPos + 1
+        && endPos < src.length()
+        && "*+?{".indexOf(src.charAt(endPos)) != -1) {
+      endPos--; // Last constant belongs under an upcoming quantifier.
```

Comment:
Does this deal correctly with non-BMP characters?

Reply:
Hm, interesting question and it touches on something I forgot to bring up in the PR description.

The parsing here depends on whether the RegExp has the unicode flag `u`. If it does, they're seen as one character, otherwise two separate characters. The same issue comes up when non-BMP characters occur in a character class (in the absence of `u`, they'll be interpreted as two independent characters).

For string literals we don't know if it's going to have the unicode flag. My plan for addressing this was to parse with `u` by default (c.f. the change to parsing of character classes), and then have a query that warns about non-BMP characters that are "split up" by the parse tree at runtime (albeit not in our own AST). This query currently only exists in the test suite, as I didn't want to go through the dance of adding a new query in the same PR, and it's not a very important query, but nice to have.

In line with the handling of character classes, I'll update this to assume the `u` flag.

Reply:
Should be fixed now